expression/builtin_string.go

```diff
@@ -424,33 +424,35 @@ type lowerFunctionClass struct {
 }
 
 func (c *lowerFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
-	sig := &builtinLowerSig{newBaseBuiltinFunc(args, ctx)}
+	tp := types.NewFieldType(mysql.TypeVarchar)
+	tp.Charset, tp.Collate = charset.CharsetUTF8, charset.CollationUTF8
+	argTp := args[0].GetType()
+	if types.IsBinaryStr(argTp) || argTp.Tp == mysql.TypeNull {
+		types.SetBinChsClnFlag(tp)
+	}
+
+	bf, err := newBaseBuiltinFuncWithTp(args, tp, ctx, tpString)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	sig := &builtinLowerSig{baseStringBuiltinFunc{bf}}
 	return sig.setSelf(sig), errors.Trace(c.verifyArgs(args))
 }
 
 type builtinLowerSig struct {
-	baseBuiltinFunc
+	baseStringBuiltinFunc
 }
 
-// eval evals a builtinLowerSig.
+// evalString evals a builtinLowerSig.
 // See https://dev.mysql.com/doc/refman/5.7/en/string-functions.html#function_lower
-func (b *builtinLowerSig) eval(row []types.Datum) (d types.Datum, err error) {
-	args, err := b.evalArgs(row)
-	if err != nil {
-		return types.Datum{}, errors.Trace(err)
-	}
+func (b *builtinLowerSig) evalString(row []types.Datum) (d string, isNull bool, err error) {
+	args := b.getArgs()
 	x := args[0]
-	switch x.Kind() {
-	case types.KindNull:
-		return d, nil
-	default:
-		s, err := x.ToString()
-		if err != nil {
-			return d, errors.Trace(err)
-		}
-		d.SetString(strings.ToLower(s))
-		return d, nil
+	d, isNull, err = x.EvalString(row, b.ctx.GetSessionVars().StmtCtx)
+	if isNull || err != nil {
+		return d, isNull, errors.Trace(err)
 	}
+	return strings.ToLower(d), false, nil
```

Comment:
Please take a look at this issue [https://github.com/pingcap/tidb/issues/3659](https://github.com/pingcap/tidb/issues/3659). When the input parameter is a binary string, the result string of built-in function `UPPER` and `LOWER` is the original input string